src/Abstractions/Entities/TaskEntityContext.cs

```diff
@@ -53,11 +53,15 @@ public virtual void StartOrchestration(TaskName name, StartOrchestrationOptions
         TaskName name, object? input = null, StartOrchestrationOptions? options = null);
 
     /// <summary>
-    /// Deletes the state of this entity after the current operation completes.
+    /// Gets the current state for the entity this context is for.
     /// </summary>
-    /// <remarks>
-    /// The state deletion only takes effect after the current operation completes. Any state changes made during the
-    /// current operation will be ignored in favor of the deletion.
-    /// </remarks>
-    public abstract void DeleteState();
+    /// <param name="type">The type to retrieve the state as.</param>
+    /// <returns>The entity state.</returns>
+    public abstract object? GetState(Type type);
```

Comment:
I thought the `object?` return type was clear that if state is not present, you get `null`. I can clarify that in a comment. We already have this same behavior with `DataConverter`.

An alternative is `bool TryGetState<T>(out T? state)`, but this then becomes ambiguous in the `false` case: was there no state to begin with, or did the state not map to `T`? Throwing from here violates the expectations of the `Try*` pattern. I think will just clarify the comment for `GetState`.

Reply:
> I thought the object? return type was clear that if state is not present, you get null

The problem with that is that you get strange  behaviors. For example, for `TaskEntity<int>`: after the user assigns `this.State = 0` then `this.State == null` which looks broken (not logically sound). The reason is that assigning the default value (`0` is the default value of int) is considered deletion and therefore removes the entity state completely.

Reply:
Per our offline discussion, setting `TState` to default of a struct (`0` for an int in this example), will **not** delete state. Users will need to declare `TaskEntity<int?>` and then assign `this.State = null`. Otherwise, they can implement a `Reset` operation which performs `this.State = this.InitializeState();` - this does not "delete" state, but resets the entity to be identical to as if state was `null` as far as the user is concerned.

We will also add a built in `delete` operation for them to rely on.

Reply:
just to clarify: reset is not identical to delete as far as the user is concerned because a reset entity will still be returned by subsequent queries, while a deleted entity is not. 

Our users definitely care about this distinction. we had endless discussions about this and many issues, e.g. https://github.com/Azure/azure-functions-durable-extension/issues/931#issuecomment-563373848.
The lesson we learnt is (a) users struggle to discover how to delete entities unless we tell them very clearly how to do that, and (b) the runtime needs to reliably remove deleted entities from subsequent entity queries.